File: lazy-professor/lazy_professor.rb
class LazyProfessor
  def self.answer_key_from_answers(answers)
    answers.
      split("\n").
      map(&:chars).
      transpose.
      map {|a| a.max_by {|b| a.count(b) } }.
      join
  end
end

if __FILE__ == $0
  if ARGV.first
    puts LazyProfessor.answer_key_from_answers(File.read(ARGV.first))
    exit 0
  end

  require 'minitest/autorun'
  require 'minitest/pride'

  describe LazyProfessor do
    it "generates the answer key from sample answers" do
      sample_answers = <<-ANSWERS.delete(" ")
        ABCCADCB
        DDDCAACB
        ABDDABCB
        AADCAACC
        BBDDAACB
        ABDCCABB
        ABDDCACB
      ANSWERS

      expected = "ABDCAACB"
      actual = LazyProfessor.answer_key_from_answers(sample_answers)

      assert_equal expected, actual
    end
  end
end

File: lazy-professor/lazy_professor.hs
import System.Environment
import Data.List
import Data.Ord

getAnswers :: [String] -> IO [String]
getAnswers (x:_) = do
  l <- readFile x
  return $ lines l
getAnswers [] = do
  return [ "ABCCADCB",
           "DDDCAACB",
           "ABDDABCB",
           "AADCAACC",
           "BBDDAACB",
           "ABDCCABB",
           "ABDDCACB" ]

getAnswerKeyFromAnswers           :: [String] -> String
getAnswerKeyFromAnswers answers   = map mostCommon $ transpose answers

mostCommon    :: String -> Char
mostCommon s  = head $ maximumBy (comparing length) (group $ sort s)

main = do
  args <- getArgs
  answers <- getAnswers args

  putStrLn $ getAnswerKeyFromAnswers answers
